### DISCUSSION TITLE
503 Service unavailable - ionCube Loader - all-inkl.com

### DISCUSSION THREAD

**Discussion by ralf-e** [kimai/kimai#2859] [On-Premise: Installation & Updates] [answered]
HELP: Installation problems (web space all-inkl.com)

Hi there,
Since I had problems after the last update of Kimai (synchronization with the app, edit data) and could not narrow down or eliminate them, I have now tried a completely clean reinstallation:

- shared hosting (all-inkl.com)
    Completely new contract, new domain, no other changes / adjustments to rule out errors due to any legacy issues

- new database (serverVersion = mariadb-10.5.11)

- Kimai installed according to instructions, .env adapted (database)

- Customer and project created

!! And from here on there are problems:

When trying to edit the customer:
   "Changes could not be saved: [503] Service Unavailable"

When trying to create another customer or project:
  "Changes could not be saved: [503] Service Unavailable"
After a reload, however, these are still displayed as available.

What did I miss .. ??

**Comment by kevinpapst** [comment]
1. What is in the logfile?
2. Shared hosting is generally crap and **not** made for products like Kimai

**Comment by ralf-e** [comment]
1. no logfile, var/log / empty 
2. Before the update to 1.15.4, Kimai2 ran here for a long time (actually since Kimai2 has existed, update from v1) without problems in this environment (shared hosting all-inkl.com Premium)

So I don't understand where the sudden problem comes from. According to all-inkl, nothing has been changed ...

**Reply by kevinpapst** [reply]
Wow, okay. So "shared hosting is not always crappy" 😁 
The error "Service Unavailable" usually stands for some server issue and not the application. 
Can you install the old version of Kimai and check if the error does not happen there?
Which PHP version are you using? Did you add the database version to the connection string in .env?

**Comment by ralf-e** [comment]
PHP-Version (CLI): 7.4.23
MariaDB-Version | 10.5.11
.env:   DATABASE_URL=mysql://xxxxxxxx:xxxxxxxxxxxxxxxx@127.0.0.1:3306/xxxxxxxx?charset=utf8&serverVersion=mariadb-10.5.11

1.15.3 I'll install again today ...

**Comment by ralf-e** [comment]
with a newly installed 1.15.3 the same behavior / the same error

after a "kimai: reload", for example, creating or editing always works exactly once, then the error again on the 2nd attempt ...

I also wrote to all-inkl.com again.
According to them, nothing has been changed and the rights to the directories are legal as stated in the Kimai documentation:
Original sound all-inkl support: "The rights to the directories are set as required by the documentation.
If necessary, the manufacturer should be contacted here so that he can debug this. " 

just seen:

created a new customer

save> error message Changes could not be saved: [503] Service
Unavailable

close

+ (new customer)> at the top of the window the message "saved changes"

Reload the page, the newly created ones are now displayed despite the previous error messages

**Comment by ralf-e** [comment]
I have now imported a complete backup (database from the beginning of this month) (Kimai version 1.15.2), this works again without any problems without the above-mentioned error. 
After another update to 1.15.5. without the intermediate step via 1.15.4, the error is back again.....

**Comment by ralf-e** [comment]
After some back and forth with the all-inkl.com support, Kimai works again here, both with a completely new clean installation and with the old data. I think all-inkl.com has changed the setting after all and that also overlapped with the restore of the old backup, after which the old installation worked again (as written above).

What still doesn't work, however, is access via the Android app: "Tracker 4 Kimai 2: Unknown API error ..."

The Kimai log for this:
> [2021-10-16 19:36:34] security.DEBUG: Stored the security token in the session. {"key":"_security_secured_area"} []
[2021-10-16 19:36:34] request.CRITICAL: Uncaught PHP Exception Symfony\Component\ErrorHandler\Error\UndefinedMethodError: "Attempted to call an undefined method named "getType" of class "ReflectionProperty"." at /www/htdocs/w0104e33/kimai.enXXXXXXXXXXXX.eu/kimai2/vendor/jms/serializer/src/Metadata/Driver/TypedPropertiesDriver.php line 107 {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\UndefinedMethodError(code: 0): Attempted to call an undefined method named \"getType\" of class \"ReflectionProperty\". at /www/htdocs/w0104e33/kimai.engXXXXXXXXXXXX.eu/kimai2/vendor/jms/serializer/src/Metadata/Driver/TypedPropertiesDriver.php:107)"} []
[2021-10-16 19:36:34] php.CRITICAL: Uncaught Error: Call to undefined method ReflectionProperty::getType() {"exception":"[object] (Error(code: 0): Call to undefined method ReflectionProperty::getType() at /www/htdocs/w0104e33/kimai.enXXXXXXXXXXXX.eu/kimai2/vendor/jms/serializer/src/Metadata/Driver/TypedPropertiesDriver.php:107)"} []
[2021-10-16 19:36:34] security.DEBUG: Remember me skipped: your authenticator does not support it. {"authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.DEBUG: Guard authenticator set no success response: request continues. {"authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.INFO: Guard authentication successful! {"token":"[object] (Symfony\\Component\\Security\\Guard\\Token\\PostAuthenticationGuardToken: PostAuthenticationGuardToken(user=\"Ralf_E\", authenticated=true, roles=\"ROLE_SUPER_ADMIN, ROLE_USER\"))","authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.DEBUG: Passing guard token information to the GuardAuthenticationProvider {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.DEBUG: Calling getCredentials() on guard authenticator. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.DEBUG: Checking support on guard authenticator. {"firewall_key":"secured_area","authenticator":"App\\Security\\TokenAuthenticator"} []
[2021-10-16 19:36:34] security.DEBUG: Checking for guard authentication credentials. {"firewall_key":"secured_area","authenticators":1} []
[2021-10-16 19:36:34] request.INFO: Matched route "get_customers". {"route":"get_customers","route_parameters":{"_route":"get_customers","_controller":"App\\API\\CustomerController::cgetAction","_format":"json"},"request_uri":"https://kimai.enXXXXXXXXXX.eu/api/customers","method":"GET"} []


any tip, what could be the reason for this?

**Reply by kevinpapst** [reply]
Glad to read that it works! The error is a 100% sure sign, that you installed with PHP 7.4, but are running with PHP 7.3 (see Doctor screen). Make sure to use the same PHP version via SSH and in web.

**Comment by ralf-e** [comment]
Thank you, everything now seems to be working again.

**Reply by skipperTux** [reply]
I have the same problem, see issue #2868. What did all-inkl do to fix?
> I think all-inkl.com has changed the setting after all

What does this mean?

**Reply by ralf-e** [reply]
Tried several new installations on my old "all-inclusive premium" website, on which Kimai has been running without problems since Kimai2 has existed, and on a new "100% clean" account.
Kimai didn't work properly on either one.
Then wrote back and forth with all-inclusive support, resetting account, ...., according to all-inkl.com everything was okay. ..... again written there, whether there were any problems with PHP / Composer, .... suddenly it worked again. First on the new account and then after a new installation + data import also on the old one.

But be careful that the PHP version of web space and CLI / SSH are identical. With the new account this was by default PHP 7.3 (Web) and PHP 7.4 (CLI) 

@kevinpapst  possibly the problem with different PHP versions should be mentioned in the Kimai documentation (installation & update)

**Reply by kevinpapst** [reply]
And I have another customer at all-ink with the exact same problem. @ralf-e can you share your ticket number or something else, so we can reference your discussion when contacting all-ink?

**Reply by ralf-e** [reply]
As already written above:
alle-inkl.com could allegedly not find any errors. But has reset the SSH user twice
(Tiket was: [NMM: 156360F])

Several installation attempts after that did not work either (neither under my old nor a new account / contract) and always returned the above errors.

But suddenly it worked.

Kimai reinstalled and tested, worked again.

Old data from .csv newly imported (.csv adapted: header changed to the English terms (was exported in German), column "Hourly rate" converted "EUR 50.00"> 50.00, some ancient data records that were incorrect at first glance ( completely wrong field entries)
removed)

Everything is fine again ..

I just tested the whole thing again with a new account, it worked immediately.
my steps exactly: 

New all-ink.com account (Premium) created
Subdomain created, SSL activated (Let's Encrypt)
via SSH to the subdomain directory

$git clone -b 1.15.5 --depth 1 https://github.com/kevinpapst/kimai2.git
$cd kimai2/
$php73 /usr/bin/composer install --no-dev --optimize-autoloader

Database created and entered in .env MariaDB-Version 10.5.11):

DATABASE_URL=mysql://USER:PASSWORD@127.0.0.1:3306/USER?charset=utf8&serverVersion=mariadb-10.5.1

$php73 bin/console kimai:install -n

$php73 bin/console kimai:user:create USER USER@XXXXXXX.DE ROLE_SUPER_ADMIN

KAS: subdomain-Target: /kimaitest.engelhardt-feuchtwangen.eu/kimai2/public/

**Comment by kevinpapst** [comment]
Thanks for sharing! Have you tried it with PHP 7.4 as well?

**Reply by ralf-e** [reply]
right now yes, I'm slowly getting practice installing Kimai :)

Domain and CLI: PHP 7.4.23
Installation went through without errors
Create the first customer was also still possible, when saving the second then: "Changes could not be saved: [503] Service Unavailable"

**Reply by kevinpapst** [reply]
Hahaha 😁 at last you don't have to fear a broken update from now on => you can definitely fix it in minutes.
Thanks for testing btw! Awesome, because I don't have access to an All-Inkl account.

All right, its getting warmer ... so this might be a PHP 7.4 issue on All-Ink. I am using 7.4 locally and every push is testing with 7.4 - and the error never shows up. But most hosters are running their own compiled version of PHP, so this might be an issue in their PHP environment.

Let's see if @skipperTux can confirm that a version switch fixes the issues (you need to witch the website to 7.3 and then run composer with 7.3 afterwards).

**Reply by ralf-e** [reply]
I just pushed infophp.php into my Kimai test domain, does that help?
https://kimaitest.engelhardt-feuchtwangen.eu/infophp.php

**Reply by kevinpapst** [reply]
What we have right now is a "proof of concept" that shows that this is a PHP 7.4 issue in your All-Ink account. 
If one more user could confirm that problem, you should reach out to the support again. Because until today only 3 user see the issue and all of them use All-Inkl.
Lets stop here for now and: you switch to 7.3 and we wait if someone else confirms your findings.

**Reply by skipperTux** [reply]
Sorry for being late to the party 😞.
I can confirm it is a PHP 7.4.23 issue with all-inkl. I tested an update from 1.15.2 to 1.15.6 with PHP 7.3.30, all fine, no errors. I also tested an update from 1.15.2 to 1.15.6 with PHP 8.0.10, all fine, no errors. (I wrote a Kimai install/update script in Ansible, so update is just a one-liner.) Doing the same update with PHP 7.4.23 gives the well known error `Service Unavailable`.
I will contact all-inkl support with these findings. Until they provide a solution, workaround is to use either PHP 7.3 or 8.0 with all-inkl.

**Reply by kevinpapst** [reply]
Thanks for your feedback! Please let us know what the All-Inkl support has to say.

**Reply by skipperTux** [reply]
all-inkl asked for exact URL of the HTTP request that is causing `[503] Service Unavailable`, and how the call is done (curl or else).

I will setup a demo instance of Kimai on a new all-inkl account, so Support can reproduce the error; maybe tomorrow or this weekend.

**Reply by ralf-e** [reply]
I also got the same questions from all-inclusive.
I've already given them demo access

**Comment by ralf-e** [comment]
Just received this answer from all-inkl.com support:

> Hallo Herr Engelhardt,
> 
> ich konnte das Problem reproduzieren und auch lokalisieren. Es kommt in Verbindung mit dem ioncube Loader zu stande. 
> 
> Wir haben hier - um Probleme auszuschließen - auch noch einmal die neuste Version des ioncube Loaders installiert, jedoch ohne Verbesserung. Auch habe ich einmal unsere vorherige ältere PHP 7.4 Installation reinstalliert, jedoch besteht der Fehler auch hier.
> 
> Wo genau jetzt der Verantwortliche Teil im Programmcode liegt, ist schwer zu sagen. Wir können hier fürchte ich auch ersteinmal wenig machen, außer auf Updates des Herstellers warten. Bis dahin muss ich leider empfehlen, dass sie auf eine andere PHP Version ausweichen, um das Problem so zu umschiffen. 
> 
> PS: merkwürdig ist, dass es mit deaktiviertem Ioncube Loader funktioniert und Ihre Seite läd. Wir können den Loader allerdings nicht serverweit deaktivieren, da evtl auch andere Kunden diesen nutzen. Interessant ist nur, dass zumindest Ihre Software mit als auch ohne Loader läuft und sich daher die Frage stellt, ob denn überhaupt der Loader benutzt wird?
> 
> 
> Mit freundlichen Grüßen
> Thomas Kriegel
> Supportteam

**Reply by kevinpapst** [reply]
Kannst Du diesen "ioncube Loaders" in der Konfiguration deaktivieren? Ich weiß nicht was das Ding macht, aber offenbar nicht nur gutes 😁 ansonsten ist es schön zu sehen, wie sich all-inkl da reinhängt. Hätte ich gar nicht erwartet, echt toll!

**Reply by skipperTux** [reply]
As far as I understand the website [ionCube Loader](https://www.ioncube.com/) is for encryption & licensing PHP scripts.
Looks like it is not possible to disable it for single accounts on shared hosting 😞.
[turn off ioncube loader for just one account](https://forum.ioncube.com/viewtopic.php?p=786&sid=e6b66a43dc72074aeb82dac8c90f26db)

**Reply by kevinpapst** [reply]
Quote from the forum: 

> In the unlikely event that you do have any possible Loader related issues, simply create a ticket in the helpdesk and we'll be pleased to assist

Someone should do that^^

**Comment by kevinpapst** [comment]
Just to share a link with everyone: https://github.com/symfony/symfony/discussions/44270
There are more applications involved than just Kimai...

**Comment by juup** [comment, marked as answer]
As https://github.com/kimai/kimai/issues/3940 is already locked, here is a solution from all-inkl support:
Create two files 
- _/www/htdocs/w00000aa/kimai/.user.ini_
-  _/www/htdocs/w00000aa/kimai/public/.user.ini_ 

Bot should contain one line:
```
ioncube.loader.encoded_paths = /www/htdocs/w00000aa/logs
```

Seems to work for Kimai 2.0.9 with PHP 8.1

**Reply by kevinpapst** [reply]
Did they really find a solution (or better a workaround) ?  That is good news 👍 
I hope someone else con confirm that, then I will add it to the documentation.

**Reply by juup** [reply]
For my system it looks like it's enough to have just 
_/www/htdocs/w00000aa/kimai/public/.user.ini_, 
But this is what they wrote as solution. 
First solution was only
/www/htdocs/w00000aa/kimai/.user.ini
My system worked better, but still had some trouble. So they told to add to _public_, too.

**Reply by bastfincke** [reply]
Thanks, it worked for me, too!

**Reply by kevinpapst** [reply]
I can confirm as well: it works 🚀  and only the file `kimai/public/.user.ini` is needed.

**Comment by whoevelmanns** [comment]
I had some trobles in starting any side twice in kimai, but with the solution with .user.ini in the public directory all problems gone.
I have changed the content to ioncube.loader.encoded_paths = /www/htdocs/{my_directory}/logs

**Reply by kevinpapst** [reply]
Thanks for your feedback. I added the workaround to the installation docs, as it seems to be necessary for All-Inkl.
